File: src/cpp/src/monte_carlo.cc
#include "monte_carlo.hpp"
#include <cmath>
#include <omp.h>

namespace mc {

MonteCarloEngine::MonteCarloEngine(size_t n_paths, size_t n_steps, unsigned int seed)
    : n_paths_(n_paths), n_steps_(n_steps), rng_(seed) {}

void MonteCarloEngine::generate_path(std::vector<double>& path, const OptionParameters& params) {
    path.resize(n_steps_ + 1);
    path[0] = params.S0;
    
    double dt = params.T / n_steps_;
    double drift = (params.r - 0.5 * params.sigma * params.sigma) * dt;
    double vol_sqrt_dt = params.sigma * std::sqrt(dt);
    
    for (size_t i = 1; i <= n_steps_; ++i) {
        double z = rng_.normal();
        path[i] = path[i-1] * std::exp(drift + vol_sqrt_dt * z);
    }
}

double MonteCarloEngine::simulate_single_path(const Option& option) {
    std::vector<double> path;
    generate_path(path, option.params());
    return option.payoff(path);
}

PricingResult MonteCarloEngine::price(const Option& option, bool use_antithetic) {
    std::vector<double> payoffs(n_paths_);
    
    if (use_antithetic) {
        // Use antithetic variates
        size_t half_paths = n_paths_ / 2;
        
        #pragma omp parallel
        {
            RandomGenerator local_rng(rng_.uniform() * 1000000);
            std::vector<double> path(n_steps_ + 1);
            std::vector<double> anti_path(n_steps_ + 1);
            
            #pragma omp for
            for (size_t i = 0; i < half_paths; ++i) {
                // Generate path
                path[0] = option.params().S0;
                anti_path[0] = option.params().S0;
                
                double dt = option.params().T / n_steps_;
                double drift = (option.params().r - 0.5 * option.params().sigma * option.params().sigma) * dt;
                double vol_sqrt_dt = option.params().sigma * std::sqrt(dt);
                
                for (size_t j = 1; j <= n_steps_; ++j) {
                    double z = local_rng.normal();
                    path[j] = path[j-1] * std::exp(drift + vol_sqrt_dt * z);
                    anti_path[j] = anti_path[j-1] * std::exp(drift - vol_sqrt_dt * z);
                }
                
                payoffs[2*i] = option.payoff(path);
                payoffs[2*i + 1] = option.payoff(anti_path);
            }
        }
    } else {
        // Standard Monte Carlo
        #pragma omp parallel
        {
            RandomGenerator local_rng(rng_.uniform() * 1000000);
            std::vector<double> path;
            
            #pragma omp for
            for (size_t i = 0; i < n_paths_; ++i) {
                path.resize(n_steps_ + 1);
                path[0] = option.params().S0;
                
                double dt = option.params().T / n_steps_;
                double drift = (option.params().r - 0.5 * option.params().sigma * option.params().sigma) * dt;
                double vol_sqrt_dt = option.params().sigma * std::sqrt(dt);
                
                for (size_t j = 1; j <= n_steps_; ++j) {
                    double z = local_rng.normal();
                    path[j] = path[j-1] * std::exp(drift + vol_sqrt_dt * z);
                }
                
                payoffs[i] = option.payoff(path);
            }
        }
    }
    
    // Calculate statistics
    double sum = 0.0;
    double sum_sq = 0.0;
    
    for (double p : payoffs) {
        sum += p;
        sum_sq += p * p;
    }
    
    double mean = sum / n_paths_;
    double variance = (sum_sq / n_paths_) - (mean * mean);
    double std_error = std::sqrt(variance / n_paths_);
    
    // Discount to present value
    double discount = std::exp(-option.params().r * option.params().T);
    double price = mean * discount;
    double se = std_error * discount;
    
    // 95% confidence interval
    double z_score = 1.96;
    PricingResult result;
    result.price = price;
    result.std_error = se;
    result.confidence_lower = price - z_score * se;
    result.confidence_upper = price + z_score * se;
    
    return result;
}

PricingResult MonteCarloEngine::price_with_control_variate(
    const Option& option,
    const EuropeanOption& control_option) {
    
    std::vector<double> target_payoffs(n_paths_);
    std::vector<double> control_payoffs(n_paths_);
    
    #pragma omp parallel
    {
        RandomGenerator local_rng(rng_.uniform() * 1000000);
        std::vector<double> path;
        
        #pragma omp for
        for (size_t i = 0; i < n_paths_; ++i) {
            path.resize(n_steps_ + 1);
            path[0] = option.params().S0;
            
            double dt = option.params().T / n_steps_;
            double drift = (option.params().r - 0.5 * option.params().sigma * option.params().sigma) * dt;
            double vol_sqrt_dt = option.params().sigma * std::sqrt(dt);
            
            for (size_t j = 1; j <= n_steps_; ++j) {
                double z = local_rng.normal();
                path[j] = path[j-1] * std::exp(drift + vol_sqrt_dt * z);
            }
            
            target_payoffs[i] = option.payoff(path);
            control_payoffs[i] = control_option.payoff(path);
        }
    }
    
    // Calculate correlation and optimal c
    double mean_target = 0.0, mean_control = 0.0;
    for (size_t i = 0; i < n_paths_; ++i) {
        mean_target += target_payoffs[i];
        mean_control += control_payoffs[i];
    }
    mean_target /= n_paths_;
    mean_control /= n_paths_;
    
    double cov = 0.0, var_control = 0.0;
    for (size_t i = 0; i < n_paths_; ++i) {
        double dt = target_payoffs[i] - mean_target;
        double dc = control_payoffs[i] - mean_control;
        cov += dt * dc;
        var_control += dc * dc;
    }
    
    double c = cov / (var_control + EPSILON);
    double control_exact = control_option.black_scholes_price() / std::exp(-option.params().r * option.params().T);
    
    // Apply control variate
    std::vector<double> adjusted_payoffs(n_paths_);
    for (size_t i = 0; i < n_paths_; ++i) {
        adjusted_payoffs[i] = target_payoffs[i] - c * (control_payoffs[i] - control_exact);
    }
    
    // Calculate statistics
    double sum = 0.0, sum_sq = 0.0;
    for (double p : adjusted_payoffs) {
        sum += p;
        sum_sq += p * p;
    }
    
    double mean = sum / n_paths_;
    double variance = (sum_sq / n_paths_) - (mean * mean);
    double std_error = std::sqrt(variance / n_paths_);
    
    double discount = std::exp(-option.params().r * option.params().T);
    
    PricingResult result;
    result.price = mean * discount;
    result.std_error = std_error * discount;
    result.confidence_lower = result.price - 1.96 * result.std_error;
    result.confidence_upper = result.price + 1.96 * result.std_error;
    
    return result;
}

} // namespace mc

File: src/python/monte_carlo.py
"""
Monte Carlo simulation engine for option pricing.

Implements geometric Brownian motion path generation and
option pricing with variance reduction techniques.
"""

from typing import Optional, Tuple
import numpy as np
from .options import Option


class MonteCarloEngine:
    """
    Monte Carlo simulation engine for pricing options.
    
    Uses geometric Brownian motion to simulate stock price paths
    and prices options by averaging discounted payoffs.
    """
    
    def __init__(self, n_paths: int, n_steps: int, seed: int = 42):
        """
        Initialize Monte Carlo engine.
        
        Args:
            n_paths: Number of simulation paths
            n_steps: Number of time steps per path
            seed: Random seed for reproducibility
        """
        if n_paths <= 0:
            raise ValueError(f"n_paths must be positive, got {n_paths}")
        if n_steps <= 0:
            raise ValueError(f"n_steps must be positive, got {n_steps}")
        
        self._n_paths = n_paths
        self._n_steps = n_steps
        self._seed = seed
        self._rng = np.random.default_rng(seed)
    
    @property
    def n_paths(self) -> int:
        """Get number of simulation paths."""
        return self._n_paths
    
    @property
    def n_steps(self) -> int:
        """Get number of time steps."""
        return self._n_steps
    
    @property
    def seed(self) -> int:
        """Get random seed."""
        return self._seed
    
    def set_n_paths(self, n_paths: int) -> None:
        """Set number of simulation paths."""
        if n_paths <= 0:
            raise ValueError(f"n_paths must be positive, got {n_paths}")
        self._n_paths = n_paths
    
    def set_n_steps(self, n_steps: int) -> None:
        """Set number of time steps."""
        if n_steps <= 0:
            raise ValueError(f"n_steps must be positive, got {n_steps}")
        self._n_steps = n_steps
    
    def set_seed(self, seed: int) -> None:
        """Set random seed and reset RNG."""
        self._seed = seed
        self._rng = np.random.default_rng(seed)
    
    def generate_paths(
        self,
        S0: float,
        r: float,
        sigma: float,
        T: float,
        antithetic: bool = False
    ) -> np.ndarray:
        """
        Generate stock price paths using geometric Brownian motion.
        
        The stochastic differential equation:
            dS = r*S*dt + sigma*S*dW
        
        Solution:
            S(t+dt) = S(t) * exp((r - 0.5*sigma^2)*dt + sigma*sqrt(dt)*Z)
        
        Args:
            S0: Initial stock price
            r: Risk-free rate (annualized)
            sigma: Volatility (annualized)
            T: Time to maturity in years
            antithetic: If True, use antithetic variates for variance reduction
        
        Returns:
            Array of price paths, shape (n_paths, n_steps+1)
            First column is S0 for all paths
        """
        dt = T / self._n_steps
        
        if antithetic:
            # Generate half the paths, then use antithetic variates for the other half
            n_half = self._n_paths // 2
            n_actual = n_half * 2  # Ensure even number
            
            # Generate random normal samples for half the paths
            Z = self._rng.standard_normal((n_half, self._n_steps))
            
            # Create antithetic pairs: [Z, -Z]
            Z_full = np.vstack([Z, -Z])
        else:
            # Standard random number generation
            Z_full = self._rng.standard_normal((self._n_paths, self._n_steps))
            n_actual = self._n_paths
        
        # Initialize paths array with S0
        paths = np.zeros((n_actual, self._n_steps + 1))
        paths[:, 0] = S0
        
        # Calculate drift and diffusion components
        drift = (r - 0.5 * sigma ** 2) * dt
        diffusion = sigma * np. sqrt(dt)
        
        # Generate paths using cumulative product
        # log(S(t+dt)/S(t)) = drift + diffusion * Z
        log_returns = drift + diffusion * Z_full
        
        # Cumulative sum of log returns gives log(S(t)/S0)
        cumulative_returns = np.cumsum(log_returns, axis=1)
        
        # S(t) = S0 * exp(cumulative_returns)
        paths[:, 1:] = S0 * np. exp(cumulative_returns)
        
        return paths
    
    def price_option(
        self,
        option: Option,
        antithetic: bool = False
    ) -> Tuple[float, float]:
        """
        Price an option using Monte Carlo simulation. 
        
        Args:
            option: Option object to price
            antithetic: If True, use antithetic variates
        
        Returns:
            Tuple of (price, standard_error)
            - price: Estimated option price
            - standard_error: Standard error of the estimate (95% CI width)
        """
        # Generate price paths
        paths = self.generate_paths(
            S0=option.params.S0,
            r=option.params.r,
            sigma=option.params.sigma,
            T=option.params.T,
            antithetic=antithetic
        )
        
        # Calculate payoffs
        payoffs = option. payoff(paths)
        
        # Discount to present value
        discount_factor = np.exp(-option.params. r * option.params.T)
        discounted_payoffs = payoffs * discount_factor
        
        # Calculate price as mean of discounted payoffs
        price = np.mean(discounted_payoffs)
        
        # Calculate standard error for 95% confidence interval
        # SE = std / sqrt(n), 95% CI ≈ 1.96 * SE
        std_dev = np.std(discounted_payoffs, ddof=1)
        n_actual = len(discounted_payoffs)
        standard_error = 1.96 * std_dev / np.sqrt(n_actual)
        
        return price, standard_error
